File: frontend/src/api/notes.ts
import { apiClient, ApiEnvelope } from "./client";

export type NoteType = "lesson_note" | "revision_note" | "cheat_sheet" | "summary";
export type NoteSource = "ai_generated" | "user_written";

export interface NoteResponse {
  id: string;
  profile_id: string;
  roadmap_node_id?: string | null;
  title: string;
  content: string;
  note_type: NoteType;
  source: NoteSource;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface NoteSummary {
  id: string;
  profile_id: string;
  roadmap_node_id?: string | null;
  title: string;
  note_type: NoteType;
  source: NoteSource;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export const notesApi = {
  generateNote: async (
    profileId: string,
    payload: {
      topic_title?: string;
      roadmap_node_id?: string;
      note_type?: NoteType;
      custom_instructions?: string;
      document_ids?: string[];
    }
  ): Promise<NoteResponse> => {
    const res = await apiClient.post<ApiEnvelope<NoteResponse>>(
      `/profiles/${profileId}/notes/generate`,
      payload
    );
    return res.data.data;
  },

  listNotes: async (
    profileId: string,
    params?: { note_type?: string; roadmap_node_id?: string; q?: string }
  ): Promise<NoteSummary[]> => {
    const res = await apiClient.get<ApiEnvelope<NoteSummary[]>>(
      `/profiles/${profileId}/notes`,
      { params }
    );
    return res.data.data;
  },

  getNote: async (profileId: string, noteId: string): Promise<NoteResponse> => {
    const res = await apiClient.get<ApiEnvelope<NoteResponse>>(
      `/profiles/${profileId}/notes/${noteId}`
    );
    return res.data.data;
  },

  createUserNote: async (
    profileId: string,
    payload: {
      title: string;
      content: string;
      note_type?: NoteType;
      roadmap_node_id?: string;
      tags?: string[];
    }
  ): Promise<NoteResponse> => {
    const res = await apiClient.post<ApiEnvelope<NoteResponse>>(
      `/profiles/${profileId}/notes`,
      payload
    );
    return res.data.data;
  },

  updateNote: async (
    profileId: string,
    noteId: string,
    payload: {
      title?: string;
      content?: string;
      note_type?: NoteType;
      tags?: string[];
    }
  ): Promise<NoteResponse> => {
    const res = await apiClient.patch<ApiEnvelope<NoteResponse>>(
      `/profiles/${profileId}/notes/${noteId}`,
      payload
    );
    return res.data.data;
  },

  deleteNote: async (profileId: string, noteId: string): Promise<boolean> => {
    const res = await apiClient.delete<ApiEnvelope<{ deleted: boolean }>>(
      `/profiles/${profileId}/notes/${noteId}`
    );
    return res.data.data.deleted;
  },
};

File: backend/app/schemas/notes.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field

NoteType = Literal["lesson_note", "revision_note", "cheat_sheet", "summary"]
NoteSource = Literal["ai_generated", "user_written"]


class NoteCreate(BaseModel):
    title: str = Field(..., min_length=1, description="Note title")
    content: str = Field(..., min_length=1, description="Markdown body content")
    note_type: NoteType = Field("lesson_note", description="Note modality")
    roadmap_node_id: str | None = Field(None, description="Optional roadmap node ID link")
    tags: list[str] = Field(default_factory=list, description="Tag list")


class NoteUpdate(BaseModel):
    title: str | None = None
    content: str | None = None
    note_type: NoteType | None = None
    tags: list[str] | None = None


class NoteGenerateRequest(BaseModel):
    topic_title: str | None = Field(None, description="Topic or concept to generate note for")
    roadmap_node_id: str | None = Field(None, description="Optional roadmap node ID")
    note_type: NoteType = Field("lesson_note", description="lesson_note, revision_note, cheat_sheet, summary")
    custom_instructions: str | None = Field(None, description="Additional custom guidance or focus areas")
    document_ids: list[str] | None = Field(None, description="Specific document IDs to ground the note in")


class NoteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    profile_id: str
    roadmap_node_id: str | None = None
    title: str
    content: str
    note_type: NoteType
    source: NoteSource
    tags: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class NoteSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    profile_id: str
    roadmap_node_id: str | None = None
    title: str
    note_type: NoteType
    source: NoteSource
    tags: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
